tests/Collection/CodecCollectionFunctionalTest.php

```diff
@@ -0,0 +1,530 @@
+            'expected' => TestObject::createDecodedForFixture(1),
+            'options' => [],
+        ];
+
+        yield 'No codec' => [
+            'expected' => self::createFixtureResult(1),
+            'options' => ['codec' => null],
+        ];
+
+        yield 'BSON type map' => [
+            'expected' => Document::fromPHP(self::createFixtureResult(1)),
+            'options' => ['typeMap' => ['root' => 'bson']],
+        ];
+    }
+
+    /** @dataProvider provideFindOneOptions */
+    public function testFindOne($expected, $options): void
+    {
+        $this->createFixtures(1);
+
+        $document = $this->collection->findOne([], $options);
+
+        $this->assertEquals($expected, $document);
+    }
+
+    public function testFindOneWithCodecAndTypemap(): void
+    {
+        $options = [
+            'codec' => new TestDocumentCodec(),
+            'typeMap' => ['root' => 'array', 'document' => 'array'],
+        ];
+
+        $this->expectExceptionObject(InvalidArgumentException::cannotCombineCodecAndTypeMap());
+        $this->collection->findOne([], $options);
+    }
+
+    public static function provideInsertManyOptions(): Generator
+    {
+        yield 'Default codec' => [
+            'expected' => [
+                TestObject::createDecodedForFixture(1),
+                TestObject::createDecodedForFixture(2),
+                TestObject::createDecodedForFixture(3),
+            ],
+            'options' => [],
+        ];
+
+        yield 'No codec' => [
+            'expected' => [
+                self::createObjectFixtureResult(1, true),
+                self::createObjectFixtureResult(2, true),
+                self::createObjectFixtureResult(3, true),
+            ],
+            'options' => ['codec' => null],
+        ];
+
+        yield 'BSON type map' => [
+            'expected' => [
+                Document::fromPHP(self::createObjectFixtureResult(1, true)),
+                Document::fromPHP(self::createObjectFixtureResult(2, true)),
+                Document::fromPHP(self::createObjectFixtureResult(3, true)),
+            ],
+            'options' => ['typeMap' => ['root' => 'bson']],
+        ];
+    }
+
+    /** @dataProvider provideInsertManyOptions */
+    public function testInsertMany($expected, $options): void
+    {
+        $documents = [
+            TestObject::createForFixture(1),
+            TestObject::createForFixture(2),
+            TestObject::createForFixture(3),
+        ];
+
+        $result = $this->collection->insertMany($documents, $options);
+        $this->assertSame(3, $result->getInsertedCount());
+
+        foreach ($expected as $index => &$expectedDocument) {
+            if ($expectedDocument instanceof BSONDocument && $expectedDocument->_id === null) {
+                $expectedDocument->_id = $result->getInsertedIds()[$index];
+            }
+
+            if ($expectedDocument instanceof Document && $expectedDocument->get('_id') === null) {
```

Comment:
Same question as below about whether these code paths are entered. Do any of the expected documents have a `null` ID?

Reply:
I'll resolve this as you already answered it in https://github.com/mongodb/mongo-php-library/pull/1140#discussion_r1295736142, but do not my follow-up comment there.